renku/core/project/project_properties.py

```diff
@@ -17,68 +17,364 @@
+    save_changes: bool = False
+    transaction_id: Optional[str] = None
 
 
 class ProjectProperties(threading.local):
     """A configuration class to hold global configuration."""
 
+    external_storage_requested = True
+    """External storage (e.g. LFS) requested for Renku command."""
+
     def __init__(self) -> None:
-        path = default_path()
+        self._context_stack: List[ProjectContext] = []
 
-        path = Path(path).resolve()
+    def __del__(self):
+        self.clear()
+
+    @property
+    def database(self) -> "Database":
+        """Current database."""
+        if not self._top.database:
+            from renku.infrastructure.database import Database
+
+            self._top.database = Database.from_path(self.database_path)
+
+        return self._top.database
+
+    @property
+    def database_path(self) -> Path:
+        """Path to the metadata storage directory."""
+        return self.metadata_path / DATABASE_PATH
 
-        self._path_stack = [path]
+    @property
+    def datadir(self) -> str:
+        """Define a name of the folder for storing datasets."""
+        from renku.core.config import get_value
+        from renku.domain_model.enums import ConfigFilter
+
+        if not self._top.datadir:
+            datadir = get_value("renku", DATA_DIR_CONFIG_KEY, config_filter=ConfigFilter.LOCAL_ONLY)
+            self._top.datadir = datadir or DEFAULT_DATA_DIR
+
+        return self._top.datadir
+
+    @datadir.setter
+    def datadir(self, value: str):
+        """Set the current datadir."""
+        self._top.datadir = value
+
+    @property
+    def dataset_images_path(self) -> Path:
+        """Return a ``Path`` instance of Renku dataset metadata folder."""
+        return self.path / RENKU_HOME / DATASET_IMAGES
 
     @property
-    def path(self):
+    def docker_path(self):
+        """Path to the Dockerfile."""
+        return self.path / DOCKERFILE
+
+    @property
+    def empty(self) -> bool:
+        """Return True if there is no pushed context."""
+        return not bool(self._context_stack)
+
+    @property
+    def global_config_dir(self) -> str:
+        """Return user's config directory."""
+        import click
+
+        return click.get_app_dir(APP_NAME, force_posix=True)
+
+    @property
+    def global_config_path(self) -> Path:
+        """Renku global (user's) config path."""
+        config = Path(self.global_config_dir)
+        if not config.exists():
+            config.mkdir(parents=True)
+
+        return config / CONFIG_NAME
+
+    @property
+    def global_config_read_lock(self):
+        """Create a user-level config read lock."""
+        from renku.core.util.contexts import Lock
+
+        return Lock(self.global_config_path)
+
+    @property
+    def global_config_write_lock(self):
+        """Create a user-level config write lock."""
+        from renku.core.util.contexts import Lock
+
+        return Lock(self.global_config_path, mode="exclusive")
```

Comment:
I moved them to `core/config` but I left them as functions; I feel it looks Javaish if we include classes with only static methods since we can have functions in Python.